### PR TITLE
`bundle inject` with source and group options

### PR DESCRIPTION
**Pull request by Shekharrajak**
Fixes https://github.com/bundler/bundler/issues/5452

Eg

```
$ bundle inject "bootstrap" ">0" --source=https://rubygems.org --group=development
Fetching gem metadata from https://rubygems.org/............
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/.
Fetching gem metadata from https://rubygems.org/.............
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/.
Added to Gemfile:
  bootstrap (> 0), group => [:development], :source => 'https://rubygems.org'
```

In GemFile 

```
gem 'bootstrap', '> 0', :group => [:development], :source => 'https://rubygems.org'
```

### Multiple group : 

```
$ dbundle inject "bootstrap" ">0" --source=https://rubygems.org --group=development,production
Fetching gem metadata from https://rubygems.org/............

Added to Gemfile:
gem 'bootstrap', '> 0', :group => [:development, :production], :source => 'https://rubygems.org'

```
In gemfile

```
# Added at 2017-03-24 11:40:51 +0530 by shekharrajak:
gem 'bootstrap', '> 0', :group => [:development, :production], :source => 'https://rubygems.org'
```

### CODE REVIEW

File: spec/commands/inject_spec.rb [hunk 1]
```diff
@@ -52,6 +52,26 @@
     end
   end
 
+  context "use source and group options" do
+    it "add gem with source in gemfile"  do
+      bundle "inject 'bootstrap' '>0' --source=https://rubygems.org"
+      gemfile = bundled_app("Gemfile").read
+      expect(gemfile.scan(/gem "bootstrap", "> 0", source => "https://rubygems.org"/).size).to eq(1)
```
Comment (Shekharrajak) [inline spec/commands/inject_spec.rb]: I have wrote these wrongly. I will correct these soon.

File: lib/bundler/cli/inject.rb [hunk 2]
```diff
@@ -31,7 +31,7 @@ def run
 
       if added.any?
         Bundler.ui.confirm "Added to Gemfile:"
-        Bundler.ui.confirm added.map {|g| "  #{g}" }.join("\n")
+        Bundler.ui.confirm added.map {|g| "  #{g}, group => #{g.groups.inspect}, :source => '#{g.source}'" }.join("\n")
```
Comment (segiddins) [inline lib/bundler/cli/inject.rb]: This should probably use `source.dump`

File: spec/commands/inject_spec.rb [hunk 3]
```diff
@@ -52,6 +52,29 @@
     end
   end
 
+  context "use source and group options" do
```
Comment (colby-swandale) [inline spec/commands/inject_spec.rb]: Can you please separate the `source` and `group` options into separate contexts

File: spec/commands/inject_spec.rb [hunk 4]
```diff
@@ -52,6 +52,29 @@
...
+    it "add gem with source in gemfile" do
+      bundle "inject 'bootstrap' '>0' --source=https://rubygems.org"
+      gemfile = bundled_app("Gemfile").read
+      str = "gem 'bootstrap', '> 0', :source => 'https://rubygems.org'"
+      expect(gemfile).to match(/str/)
+    end
```
Comment (colby-swandale) [inline spec/commands/inject_spec.rb]: I prefer to use `include` instead if were not searching with a regex. Example:

    expect(gemfile).to include "gem 'bootstrap', '> 0', :source => 'https://rubygems.org'"

Same with the other expectations below as well.

File: lib/bundler/cli.rb [hunk 5]
```diff
@@ -465,10 +465,14 @@ def platform
     end
 
     desc "inject GEM VERSION", "Add the named gem, with version requirements, to the resolved Gemfile"
+    method_option "source", :type => :string, :banner =>
```
Comment (colby-swandale) [inline lib/bundler/cli.rb]: Maybe `"install gem from the given source"` and `"Install gem into a bundler group"`

File: lib/bundler/cli/inject.rb (hunk 2 above)
Comment (Shekharrajak) [inline lib/bundler/cli/inject.rb]: I didn't understand what this means ? @segiddins,  Can you tell me what should be replaced.

File: lib/bundler/cli.rb (hunk 5 above)
Comment (Shekharrajak) [inline lib/bundler/cli.rb]: Thanks for review @colby-swandale . I have edited the PR according to to your comments.

**Comment by Shekharrajak** [comment]
I didn't get why I am getting https://travis-ci.org/bundler/bundler/jobs/203772974#L425  error in test, while it is working fine when we run this  
`$ dbundle inject "bootstrap" ">0" --source=https://rubygems.org --group=development` in terminal. 

I think this is the main error `Could not reach host index.rubygems.org. Check your network connection and try again.` But it must go to `rubygems.org` (not `index.rubygems.org`)

**Comment by Shekharrajak** [comment]
I have done the required changes and travis build passed as well. Please review it now.

**Comment by Shekharrajak** [comment]
Ping @colby-swandale @segiddins @indirect , please review.

**Comment by colby-swandale** [comment]
Please be patient, we will get around to reviewing the PR when we can.

File: lib/bundler/cli/inject.rb [hunk 6]
```diff
@@ -31,7 +31,7 @@ def run
 
       if added.any?
         Bundler.ui.confirm "Added to Gemfile:"
-        Bundler.ui.confirm added.map {|g| "  #{g}" }.join("\n")
+        Bundler.ui.confirm added.map {|g| "  #{g}, group => #{g.groups.inspect}, :source => '#{g.source.dump}'" }.join("\n")
```
Comment (segiddins) [inline lib/bundler/cli/inject.rb]: should only include the groups and the source if they're passed in

File: spec/commands/inject_spec.rb [hunk 7]
```diff
@@ -52,6 +52,24 @@
...
+    it "add gem with source option in gemfile" do
+      bundle "inject 'foo' '>0' --source file://#{gem_repo1}"
+      gemfile = bundled_app("Gemfile").read
+      str = "gem 'foo', '> 0', :source => 'file://#{gem_repo1}'"
+      expect(gemfile).to include str
+    end
+  end
+
+  context "with group option" do
+    it "add gem with group option in gemfile" do
```
Comment (segiddins) [inline spec/commands/inject_spec.rb]: please also add a test for including multiple groups

**Comment by colby-swandale** [comment]
ping @Shekharrajak

**Comment by Shekharrajak** [comment]
Sorry I completely missed the notification.  I will edit the PR accordingly soon.

File: spec/commands/inject_spec.rb (hunk 7 above)
Comment (Shekharrajak) [inline spec/commands/inject_spec.rb]: Thanks for review I have added the testcase for multiple groups in [this commit ](https://github.com/bundler/bundler/pull/5456/commits/cfe4eccfc33a8f6372a7197ffae95c15e023df95)

File: lib/bundler/cli/inject.rb (hunk 6 above)
Comment (Shekharrajak) [inline lib/bundler/cli/inject.rb]: Thanks for review. I have added new msg in [this commit](https://github.com/bundler/bundler/pull/5456/commits/cfe4eccfc33a8f6372a7197ffae95c15e023df95)

**Comment by Shekharrajak** [comment]
@colby-swandale , @segiddins Please review once when you get time . Thanks!

File: spec/commands/inject_spec.rb [hunk 8]
```diff
@@ -68,6 +68,13 @@
       str = "gem 'rack-obama', '> 0', :group => [:development]"
       expect(gemfile).to include str
     end
+
+    it "add gem with multiple group in gemfile" do
+      bundle "inject 'rack-obama' '>0' --group=development,test"
+      gemfile = bundled_app("Gemfile").read
+      str = "gem 'rack-obama', '> 0', :group => [:development, :test]"
```
Comment (segiddins) [inline spec/commands/inject_spec.rb]: shouldn't this be `groups` instead of `group`?

Comment (Shekharrajak) [inline spec/commands/inject_spec.rb]: Thanks @segiddins for review. I have edited the line  in this commit https://github.com/bundler/bundler/pull/5456/commits/1f466c04a6d3ea54abfafc4d5fde0a17717ade55

Sorry for the silly mistake.

Comment (Shekharrajak) [inline spec/commands/inject_spec.rb]: I edited this line : "add gem with multiple group in gemfile"  

and I think this line : `gem 'rack-obama', '> 0', :group => [:development, :test]` is correct. I followed this link : http://bundler.io/v1.3/groups.html  examples .

File: spec/commands/inject_spec.rb [hunk 9]
```diff
@@ -52,6 +52,31 @@
...
+      bundle "inject 'rack-obama' '>0' --group=development"
+      gemfile = bundled_app("Gemfile").read
+      str = "gem 'rack-obama', '> 0', :group => [:development]"
+      expect(gemfile).to include str
+    end
+
+    it "add gem with multiple groups in gemfile" do
+      bundle "inject 'rack-obama' '>0' --group=development,test"
+      gemfile = bundled_app("Gemfile").read
+      str = "gem 'rack-obama', '> 0', :group => [:development, :test]"
```
Comment (segiddins) [inline spec/commands/inject_spec.rb]: this still reads `group`, and given it's the key for an array it should be `groups`

Comment (Shekharrajak) [inline spec/commands/inject_spec.rb]: @segiddins , I saw example here : http://bundler.io/v1.3/groups.html . In gemfile it must be `:group => [:development, :test]` right ? I haven't see `groups ` in gemfile.  

To install all dependencies, except those in specified groups
we do something : 
```
$ bundle install --without test development
```
will it work when we write `groups` ?

Comment (segiddins) [inline spec/commands/inject_spec.rb]: it can be `groups` as well

Comment (Shekharrajak) [inline spec/commands/inject_spec.rb]: Thanks @segiddins , I have changed it in this commit https://github.com/bundler/bundler/pull/5456/commits/1754e5019fe7f97db6d24f11bba8768d1dca7772

File: lib/bundler/injector.rb [hunk 10]
```diff
@@ -51,7 +51,14 @@ def new_gem_lines
       @new_deps.map do |d|
         name = "'#{d.name}'"
         requirement = ", '#{d.requirement}'"
-        group = ", :group => #{d.groups.inspect}" if d.groups != Array(:default)
+        if d.groups != Array(:default)
+          group =
+            if d.groups.size == 1
+              ", :group => #{d.groups.inspect}"
```
Comment (segiddins) [inline lib/bundler/injector.rb]: If it's just one group, we shouldn't need to put it in an array

**Comment by segiddins** [comment]
@colby-swandale r?

**Comment by colby-swandale** [comment]
@bundlebot r+

**Comment by colby-swandale** [comment]
@bundlerbot r+

**Pull request merged**